server/battles/BattleActionProcessor.cpp

```diff
@@ -1193,35 +1258,70 @@ void BattleActionProcessor::handleAfterAttackCasting(const CBattleInfoCallback &
 		return;
 	}
 
-	if(attacker->hasBonusOfType(BonusType::DEATH_STARE))
+	if(attacker->hasBonusOfType(BonusType::DEATH_STARE) || attacker->hasBonusOfType(BonusType::ACCURATE_SHOT))
 	{
 		// mechanics of Death Stare as in H3:
 		// each gorgon have 10% chance to kill (counted separately in H3) -> binomial distribution
 		//original formula x = min(x, (gorgons_count + 9)/10);
 
-		double chanceToKill = attacker->valOfBonuses(BonusType::DEATH_STARE, BonusCustomSubtype::deathStareGorgon) / 100.0f;
-		vstd::amin(chanceToKill, 1); //cap at 100%
+		/* mechanics of Accurate Shot as in HotA:
+		 * each creature in an attacking stack has a X% chance of killing a creature in the attacked squad,
+		 * but the total number of killed creatures cannot be more than (number of creatures in an attacking squad) * X/100 (rounded up).
+		 * X = 3 multiplier for shooting without penalty and X = 2 if shooting with penalty. Ability doesn't work if shooting at creatures behind walls.
+		*/
+
+		auto bonus = attacker->getBonus(Selector::type()(BonusType::DEATH_STARE));
```

Comment:
please at the very least extract this... code to separate method

Reply:
You may also want to check 4th commit and tell me if you prefer split version, though it got bunch of code duplication

Reply:
Move this version to a separate method